### DISCUSSION TITLE
(Feature request) Allow Expanding Notes Via API

### DISCUSSION THREAD

**Discussion by BeatLink** [TriliumNext/Trilium#7809] [Q&A]
### Describe feature

Hi. I'm trying to create a script so that notes with an #alwaysExpanded label are automatically expanded when their parents are visible in the tree. However, there doesn't seem to be an API function to expand said notes.

**Comment by BeatLink** [comment]
Having given it some thought, it would be important that this feature applies to specific branches rather than the note themselves. A user may want only some branches always expanded while others arent

**Comment by zadam** [comment]
Are you using ETAPI? Branch entity has `isExpanded` property. Setting it to `true` should expand the branch. As you say, it affects only specific branch, not the whole note.

**Comment by BeatLink** [comment]
Oh okay thanks. Ill have a look

**Comment by BeatLink** [comment]
How do i access ETAPI from within a backend script?

**Comment by zerebos** [comment]
It's just a REST API so you could use `fetch` or similar I believe, but generally if you're already in a frontend/backend script there's probably a better way to do what you're trying to do.

**Comment by BeatLink** [comment]
Essentially i want to create a script that automatically expands the children for a note with  a label like #alwaysExpanded

**Comment by zadam** [comment]
Within a backend script it's quite easy (untested, written from memory):

```javascript
const notes = api.searchForNotes('#alwaysExpanded');

for (const note of notes) {
        // expand all "placements" of note into the tree (clones)
	for (const branch of note.getParentBranches()) {
		branch.isExpanded = true;
		branch.save();
	}
}
```

Something to think about is when such a script should run.

**Comment by BeatLink** [comment]
It works, thanks! The plan right now is to have the script run whenever the parent of the notes with the given tags are expanded

**Comment by BeatLink** [comment]
...which leads me to my next question. How do you detect if a note has been expanded. I checked the built in events and the closest was runOnNoteChange which didnt work.

**Comment by BeatLink** [comment]
Also a weird bug in the script. It seems to only expand the first branch of a given note. If the note is cloned, it wont expand the others

**Comment by zadam** [comment]
There isn't such an event yet, so far nobody needed this. There are only two attributes you might want to watch for - expanded status and prefix, both of which are minor.

Having said that, I've added `runOnBranchChange` event to master, which will become 0.62

**Comment by BeatLink** [comment]
Thanks!

**Comment by BeatLink** [comment]
Hey, been trying to implement this but ive run into an issue. 

Essentially, i added the run on branch change relation to the root node and made it inheritable, which means the script can be triggered when any branch changes.

The problem is, the script is being triggered by branch changes caused by the script itself, leading to a feedback loop that freezes the GUI.

How can i fix this?

**Comment by BeatLink** [comment]
Never mind, solved it by adding a check to the script to see if the branch is already expanded and only if not, setting the expanded state. 

Eventually all of the branches reaches their desired state which stops any more branch changes and breaks the loop
